sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -462,6 +462,32 @@ object HiveTypeCoercion {
           val resultType = widerDecimalType(p1, s1, p2, s2)
           b.makeCopy(Array(Cast(e1, resultType), Cast(e2, resultType)))
 
+        // Strength reduction for comparisons between an integral column and a decimal literal:
+        // 1. int_col > decimal_literal => int_col > floor(decimal_literal)
+        // 2. int_col >= decimal_literal => int_col >= ceil(decimal_literal)
+        // 3. int_col < decimal_literal => int_col < ceil(decimal_literal)
+        // 4. int_col <= decimal_literal => int_col <= floor(decimal_literal)
+        // 5. decimal_literal > int_col => ceil(decimal_literal) > int_col
+        // 6. decimal_literal >= int_col => floor(decimal_literal) >= int_col
+        // 7. decimal_literal < int_col => floor(decimal_literal) < int_col
+        // 8. decimal_literal <= int_col => ceil(decimal_literal) <= int_col
+        case GreaterThan(i @ IntegerType(), Literal(value: Decimal, _: DecimalType)) =>
```

Comment:
this should apply for any integral type, not just integer. Also can you make this function shorter by breaking it into two functions? It is pretty long now.